test/integration/protocol_integration_test.cc

```diff
@@ -1299,14 +1299,182 @@ TEST_P(ProtocolIntegrationTest, TwoRequests) { testTwoRequests(); }
+  });
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_max_duration_reached", 1);
+
+  if (downstream_protocol_ == Http::CodecType::HTTP1) {
+    ASSERT_TRUE(codec_client_->waitForDisconnect());
+    ASSERT_TRUE(response->complete());
+  } else {
+    ASSERT_TRUE(response->waitForEndStream());
+    codec_client_->close();
+  }
+}
+
+TEST_P(ProtocolIntegrationTest, BasicDynamicMaxStreamDuration) {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  default_request_headers_.setEnvoyUpstreamStreamTimeoutMs(500);
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_max_duration_reached", 1);
+
+  if (downstream_protocol_ == Http::CodecType::HTTP1) {
+    ASSERT_TRUE(codec_client_->waitForDisconnect());
+    ASSERT_TRUE(response->complete());
+  } else {
+    ASSERT_TRUE(response->waitForEndStream());
+    codec_client_->close();
+  }
+}
+
+TEST_P(ProtocolIntegrationTest, MaxStreamDurationTimerDisarmedWithDynamicValue) {
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    ConfigHelper::HttpProtocolOptions protocol_options;
+    auto* http_protocol_options = protocol_options.mutable_common_http_protocol_options();
+    http_protocol_options->mutable_max_stream_duration()->MergeFrom(
+        ProtobufUtil::TimeUtil::MillisecondsToDuration(200));
+    ConfigHelper::setProtocolOptions(*bootstrap.mutable_static_resources()->mutable_clusters(0),
+                                     protocol_options);
+  });
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  default_request_headers_.setEnvoyUpstreamStreamTimeoutMs(0);
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_max_duration_reached", 0);
```

Comment:
I'm not quite sure I understand this test. Is the idea that the the config specifies a timeout, but the request says no timeout? If so, then the request should not timeout, right? How do we verify that it does not timeout? (I would expect that a counter would always be >= 0 since it can't be negative)

Reply:
Yes. I fixed it `waitForCounterEq` and checking not invoked max stream duration timer.

Reply:
This is just my mistake.... Fixed that.

Reply:
Ah, ok. Thanks for the fix! So it seems like the key change is this:

```

  // 1 sec is large enough to wait configured stream duration 200ms.
  absl::SleepFor(absl::Seconds(1));

  // Check that stream is not disconnected after 1 sec.
```

Is that right?

If so, I think I'd be inclined to remove this test and instead rely on the test coverage from the unit tests to cover the case where the header being set to 0 eliminates the timeout. I suggest this because in general having tests sleep is problematic. On one hand, it makes them fragile because they are sensitive to load on the test system which might make the timing unpredictable and the sleep might not be long enough. At the same time, the sleep is *usually* too long and so they end up slowing down the tests in general :( 

So, I'd remove this particular test and then I think this looks good to me!